src/resources/ble_esp32.cc

```diff
@@ -647,10 +851,27 @@ PRIMITIVE(request_data) {
     memmove(bytes.address() + offset, current->om_data, current->om_len);
     offset += current->om_len;
   }
-  gatt->set_mbuf(null);
   return data;
 }
 
+
+PRIMITIVE(request_data) {
+  ARGS(GATTResource, gatt);
+
+  if (gatt->error() == BLE_HS_ENOENT) {
+    OUT_OF_RANGE;
+  } else if (gatt->error() != 0) {
+    INVALID_ARGUMENT;
+  }
+
+  const struct os_mbuf* mbuf = gatt->mbuf();
+  if (!mbuf) return process->program()->null_object();
+  Object* ret_val = convert_mbuf_to_heap_object(process, mbuf);
+  gatt->set_mbuf(null);
```

Comment:
Do you want to do this even if the allocation of the heap object failed? That looks wrong.